feat(filtering) Use query params for filtering

aaryanporwal opened the pull request:
# Done
* Filtering by tags now also updates the search query parameters.
* This now allows user to create shareable link to a subset of masterclasses.
* Refactoring:
    * Encapsulation & Readability
    * Use of DOMContentLoaded
    * Simplified Query Selection: Reduced redundant calls to querySelectorAll by storing frequently accessed elements.
    
# QA
* Go to https://masterclasses-canonical-com-55.demos.haus/
* Try filtering using the tags "python" or "django" and check if it works as expected.
* Also open the new URL with search query params in a new tab and check if it only shows the subset of sessions. 
* Verify if the query param tag has a green (positive) chip in the filters.


[WD-11585](https://warthogs.atlassian.net/browse/WD-11585)

[WD-11585]: https://warthogs.atlassian.net/browse/WD-11585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lorumic commented on templates/masterclasses.html:
```diff
@@ -247,93 +245,100 @@ <h3 class="p-muted-heading u-sv2">Something missing?</h3>
+    };
+
+    const resetChips = () => {
+      document.querySelectorAll(".p-chip--positive").forEach((chip) => {
```
This `querySelectorAll` logic hints at a possibility to select more than one tag at a time for filtering, which would be a great user experience, but instead, this is currently not supported. Could you please change the logic below (e.g. by using `searchParams.append` instead of `searchParams.set`, and then `searchParams.getAll` instead of `searchParams.get`) so that we can allow multi-tag filtering?

aaryanporwal replied:
Following our discussion, I've fixed the usage of querySelectorAll and search reset.

aaryanporwal replied:
[Ticket](https://warthogs.atlassian.net/browse/WD-14154?atlOrigin=eyJpIjoiMDU3NjM0ODZhMzA2NDUyZTgzOTc2MmMwOTEwNWE3NmYiLCJwIjoiaiJ9) for multi-tag filtering

lorumic commented:
The search reset issue that I had mentioned yesterday on Mattermost is still there:

![Peek 2024-08-13 18-47](https://github.com/user-attachments/assets/fefb8ba3-aa22-4baa-8175-ade5eb7c87fa)

To avoid an inconsistent display of results as shown in the screen capture above, **search query and chip filtering should reset each other**. This means:

1. If a search query is there and a chip is clicked, this action should **reset the search query**.
2. If a chip is selected and a user clicks the "X" button of the search bar, this action should **reset the chip filters**. (actually, this case will never be possible when point 1 above is addressed)

Both these cases are covered and addressed in the patch that I sent you yesterday, together with the multi-item filtering implementation.

aaryanporwal commented:
> The search reset issue that I had mentioned yesterday on Mattermost is still there:
> 
> To avoid an inconsistent display of results as shown in the screen capture above, **search query and chip filtering should reset each other**. This means:
> 
> 1. If a search query is there and a chip is clicked, this action should **reset the search query**.
> 2. If a chip is selected and a user clicks the "X" button of the search bar, this action should **reset the chip filters**. (actually, this case will never be possible when point 1 above is addressed)
> 
> Both these cases are covered and addressed in the patch that I sent you yesterday, together with the multi-item filtering implementation.

Thanks for pointing that out. I've addressed the issue by resetting search when chip is clicked.

lorumic reviewed (approved):
Thanks, it now works as I would expect it to. Since we decided to opt for the limited single-item filtering I will approve, but please do wait for @edlerd's review as he might have better insights to share.

edlerd reviewed (approved):
LGTM
